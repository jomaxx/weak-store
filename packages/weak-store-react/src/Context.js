import React from "react";

export default React.createContext({
  getState: namespace => namespace.state,
  setState: () => {},
  subscribe: () => () => {}
});
