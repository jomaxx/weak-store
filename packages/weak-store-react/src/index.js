export { createWeakStore } from "weak-store";
export { Provider as WeakStoreProvider } from "./Context";
export { default as createNamespace } from "./createNamespace";
