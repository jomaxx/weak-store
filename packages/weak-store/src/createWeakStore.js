export default function createWeakStore() {
  const state = new WeakMap();
  const listeners = [];

  function getState(key) {
    // initialize key
    if (!state.has(key)) state.set(key, key.state);
    return state.get(key);
  }

  function setState(key, updater) {
    const prevState = getState(key);
    const nextState =
      typeof updater === "function" ? updater(prevState) : updater;
    if (nextState === null) return;
    state.set(key, Object.assign({}, prevState, nextState));
    listeners.forEach(listener => listener());
  }

  function subscribe(listener) {
    listeners.push(listener);

    let unsubscribed = false;

    function unsubscribe() {
      if (unsubscribed) return;
      unsubscribed = true;
      listeners.splice(listeners.indexOf(listener), 1);
    }

    return unsubscribe;
  }

  return { getState, setState, subscribe };
}
