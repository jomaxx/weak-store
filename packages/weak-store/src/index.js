export { default as createWeakStore } from "./createWeakStore";
