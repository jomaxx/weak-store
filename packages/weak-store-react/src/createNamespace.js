import React from "react";
import { Consumer } from "./Context";

export default function createNamespace(namespace = {}) {
  class Namespace extends React.PureComponent {
    state = {
      state: this.props.store.getState(namespace),
      setState: this.props.store.setState.bind(null, namespace)
    };

    componentDidMount() {
      const store = this.props.store;

      const update = () => {
        this.setState(prevState => {
          const state = store.getState(namespace);
          return state === prevState ? null : { state };
        });
      };

      this.componentWillUnmount = this.props.store.subscribe(update);

      update();
    }

    render() {
      return this.props.children(this.state);
    }
  }

  function WeakStore(props) {
    return (
      <Consumer>{store => <Namespace store={store} {...props} />}</Consumer>
    );
  }

  return WeakStore;
}
