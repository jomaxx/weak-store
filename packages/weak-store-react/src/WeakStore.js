import React from "react";
import { createWeakStore } from "weak-store";

const Context = React.createContext({
  getState: namespace => namespace.state,
  setState: () => {},
  subscribe: () => () => {}
});

function getDerivedStateFromProps(nextProps, prevState) {
  if (typeof nextProps.value !== "undefined") return { weakStore: undefined };
  if (prevState.value) return null;
  return { value: createWeakStore() };
}

export default class WeakStore extends React.Component {
  static get Consumer() {
    return Context.Consumer;
  }

  static get getDerivedStateFromProps() {
    return getDerivedStateFromProps;
  }

  state = {};

  render() {
    return (
      <Context.Provider value={this.state.value || this.props.value}>
        {this.props.children}
      </Context.Provider>
    );
  }
}
