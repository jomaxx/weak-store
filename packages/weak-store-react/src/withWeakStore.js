import React from "react";
import WeakStore from "./WeakStore";

export default function withWeakStore(C) {
  function WithWeakStore(props) {
    return (
      <WeakStore.Consumer>
        {weakStore => <C weakStore={weakStore} {...props} />}
      </WeakStore.Consumer>
    );
  }

  /* istanbul ignore next */
  WithWeakStore.displayName = `withWeakStore(${C.displayName ||
    C.name ||
    "Component"})`;

  return WithWeakStore;
}
