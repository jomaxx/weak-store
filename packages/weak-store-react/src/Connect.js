import React from "react";
import withWeakStore from "./withWeakStore";

function getDerivedStateFromProps(nextProps, prevState) {
  const nextState = {
    to: nextProps.to,
    setState: nextProps.weakStore.setState,
    value: {
      state: nextProps.weakStore.getState(nextProps.to)
    }
  };

  const hasChanges =
    !prevState.value ||
    nextState.to !== prevState.to ||
    nextState.value.state !== prevState.value.state ||
    nextState.setState !== prevState.setState;

  if (hasChanges) {
    nextState.value.setState = nextState.setState.bind(null, nextState.to);
    return nextState;
  }

  return null;
}

class Connect extends React.PureComponent {
  static get getDerivedStateFromProps() {
    return getDerivedStateFromProps;
  }

  state = {};

  componentDidMount() {
    const updater = state => getDerivedStateFromProps(this.props, state);

    this.componentWillUnmount = this.props.weakStore.subscribe(() => {
      this.setState(updater);
    });

    this.setState(updater);
  }

  componentDidUpdate(prevProps) {
    if (this.props.weakStore === prevProps.weakStore) return;
    this.componentWillUnmount();
    this.componentDidMount();
  }

  render() {
    return this.props.children(this.state.value);
  }
}

export default withWeakStore(Connect);
