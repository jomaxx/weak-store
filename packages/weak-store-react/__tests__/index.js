import React from "react";
import * as module from "../src";

test("has exports", () => {
  expect(module).toMatchSnapshot();
});
