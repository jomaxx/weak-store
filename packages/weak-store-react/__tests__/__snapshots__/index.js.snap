// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`has exports 1`] = `
Object {
  "Connect": [Function],
  "WeakStore": [Function],
  "createWeakStore": [Function],
  "withWeakStore": [Function],
}
`;
